File: src/storeFx/updatePlayQueue.ts
import { deepEqual } from '@tanstack/react-router';

import * as Fx from '../fx/fx';
import type { Store } from '../store/types';
import { savePlayQueue } from '../subsonic';

interface PlayQueueState {
  current?: string;
  currentTime: number;
  queued: string[];
}

export function make(playQueueState: (prev: PlayQueueState) => PlayQueueState) {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    const state = store.getState();

    const updatedPlayQueueState = playQueueState({
      current: state.player.currentSongId,
      currentTime: state.player.currentTime,
      queued: state.player.queuedSongIds,
    });

    if (!deepEqual(updatedPlayQueueState.queued, state.player.queuedSongIds)) {
      store.setState(prevState => ({
        albums: prevState.albums,
        artists: prevState.artists,
        auth: prevState.auth,
        player: {
          ...prevState.player,
          queuedSongIds: updatedPlayQueueState.queued,
        },
        songs: prevState.songs,
      }));
    }

    return Fx.catch_(
      Fx.provide(
        savePlayQueue(
          updatedPlayQueueState.queued,
          state.player.currentSongId ?? updatedPlayQueueState.current,
          updatedPlayQueueState.currentTime,
        ),
        { credentials: state.auth.credentials },
      ),
      () => Fx.Err(undefined),
    );
  });
}

File: src/storeFx/updatePlayQueue.ml
open Fx
open Monad_syntax

type state = {
  current : string option;
  currentTime : float;
  queued : string array;
}

let make playQueueState =
  let* Deps.{ store } = ask () in
  let state = store |. Zustand.getState in
  let playQueueState =
    playQueueState
      {
        current = state.player.currentSongId;
        currentTime = state.player.currentTime;
        queued = state.player.queuedSongIds;
      }
  in
  if not (Router.deepEqual playQueueState.queued state.player.queuedSongIds)
  then
    store
    |. Zustand.setState (fun prevState ->
        {
          prevState with
          player =
            { prevState.player with queuedSongIds = playQueueState.queued };
        });
  Subsonic.SavePlayQueue.make playQueueState.queued
    ?current:
      (state.player.currentSongId |> Belt.Option.orElse playQueueState.current)
    ~position:playQueueState.currentTime ()
  |. provide ~deps:{ credentials = state.auth.credentials }
  |. catch (fun _ -> Error ())
